kubelet apiserver: be gentle closing connections on heartbeat failures

aojea opened the pull request:
Follow up on https://github.com/kubernetes/kubernetes/pull/104844
Alternative to https://github.com/kubernetes/kubernetes/pull/107879

Kubelet was forcefully closing all connections (idle and active) on heartbeat failures https://github.com/kubernetes/kubernetes/pull/63492 
However, since https://github.com/kubernetes/kubernetes/pull/95981, all clients using HTTP2 use a health check by default that allows to detect stale connections without any additional logic.

In case users are disabling http2, by setting the environment variable DISABLE_HTTP2, the previous behavior is maintained.



/kind bug

```release-note
kubelet don't forcefully close active connections on heartbeat failures, using the http2 health check mechanism to detect broken connections. Users can force the previous behavior of the kubelet by setting the environment variable DISABLE_HTTP2.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aojea committed:
kubelet apiserver: be gentle closing connections on heartbeat failures

aojea commented:
/assign @liggitt @wojtek-t 
/cc @johnRusk

I think that this is simpler than https://github.com/kubernetes/kubernetes/pull/107879

aojea commented:
/sig api-machinery

wojtek-t commented:
This LGTM, but I would also like to hear from @liggitt

JohnRusk commented:
Might be nice to have a comment in the code to explain why the code has to close _idle_ connections on heartbeat failure.  I.e. Why does failure of a heartbeat (which is presumably happending on connection that is _not_ idle) signal to us that we need to close the idle ones?

BTW, I like the idea of relying on Pings rather than heartbeats for monitoring health of the one "live" HTTP2 connection. That looks nice.  

The bit that seems confusing to me, and I suggest may need an explainatory comment, it's just the fact that close**All**Connections gets wired up to a method that closes **Idle** connections. It's not obvious why that is necessary or correct.

aojea commented:
> The bit that seems confusing to me, and I suggest may need an explainatory comment, it's just the fact that close**All**Connections gets wired up to a method that closes **Idle** connections. It's not obvious to readers of the code (at least, not to me) why that is necessary or correct.


yeah, let's discuss the whole problem:

The thing is that the function is passed as `kubeDeps.OnHeartbeatFailure`

https://github.com/kubernetes/kubernetes/blob/d899c39ca3025362033b5a71e4f27c32690dc78b/cmd/kubelet/app/server.go#L556-L564

and then plumbed to the lease controller https://github.com/kubernetes/kubernetes/pull/107879 

So the real semantics should be  "function that we call to do things on heartbeat failures" , that previously was "closeAllConnections" and now is "closeAllIdleConnections", but it can still mean "closeAllConnections" if HTTP2 is explicitly disabled.

The current situation is that we already have retry logic in client-go on network errors (default to 10 times)

https://github.com/kubernetes/kubernetes/blob/d899c39ca3025362033b5a71e4f27c32690dc78b/staging/src/k8s.io/client-go/rest/request.go#L1000-L1012

 so some of the loops are really not needed since are multiplying the number of retries, per example, the node status update:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L451

that means that an apiserver not replying will create a max of 50 connections attempts per kubelet , because it retries 5 times

```
pkg/kubelet/kubelet.go: // nodeStatusUpdateRetry specifies how many times kubelet retries when posting node status failed.
pkg/kubelet/kubelet.go: nodeStatusUpdateRetry = 5
```

TCP sockets are expensive, but you should only notice at a relative high scale, with HTTP2 https://github.com/kubernetes/kubernetes/pull/95981 the client automatically detects idle, it has an embedded heartbeat detection, so we should not really need any heartbeat logic at all, the stdlib does it for us.
With HTTP1 you can only have one connection in the pool, so it can only be idle or active. If active the client always dial a new connection , if it is idle, we clean it to force to dial a new connection (avoid to reuse a stale connection)

This PR is the easy solution, I really don't know how to document this better, but in order to keep compatibility with old systems that still use HTTP1 I feel this is the less risky approach

fedebongio commented:
/triage accepted

JohnRusk commented:
Thanks for the explaination @aojea!  

I reckon there's no need to add any additional docs to the code, since if anyone is curious they can find this PR from the commit history in the future, and read what you wrote above.

JohnRusk commented:
This PR supercedes #107781, so I should close #107781 now, right?

matthyx commented:
> This PR supercedes #107781, so I should close #107781 now, right?

yes please :)

matthyx reviewed (approved):
/lgtm

liggitt commented on cmd/kubelet/app/server.go:
```diff
@@ -845,6 +845,17 @@ func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, nod
+		// once the Kubelet heartbeat fails. However, since there are many edge cases that we can't
+		// control, users can still opt-in to the previous behavior for closing the connections by
+		// setting the environment variable DISABLE_HTTP2.
+		if s := os.Getenv("DISABLE_HTTP2"); len(s) > 0 {
+			klog.InfoS("HTTP2 has been explicitly disabled, Kubelet will forcefully close active connections on heartbeat failures")
+		} else {
+			closeAllConns = func() { utilnet.CloseIdleConnectionsFor(transportConfig.Transport) }
+		}
```
since the function we're returning might not actually be closing all connections now, let's name the return var what it actually is (same below with the existing code, and same with the var the return value from this function is assigned to in `run()`)

```diff
+		var onHeartbeatFailure func()
		if s := os.Getenv("DISABLE_HTTP2"); len(s) > 0 {
			klog.InfoS("HTTP2 has been explicitly disabled, Kubelet will forcefully close active connections on heartbeat failures")
+			onHeartbeatFailure = closeAllConns
		} else {
-			closeAllConns = func() { utilnet.CloseIdleConnectionsFor(transportConfig.Transport) }
+			onHeartbeatFailure = func() { utilnet.CloseIdleConnectionsFor(transportConfig.Transport) }
		}
...
-		return transportConfig, closeAllConns, nil
+ 		return transportConfig, onHeartbeatFailure, nil
```

aojea replied:
added a new commit with the renames so all codebase is consistent with the name

aojea committed:
kubelet: rename closeAllConns to onHeartbeatFailure

Since the behavior of the kubelet OnHeartbeatFailure hook has changed
to close only the idle connections, rename this function to match
the semantics.

aojea commented:
E2eNode Suite: [sig-node] Summary API [NodeConformance] when querying /stats/summary should report resource usage through the stats api expand_less	


/test pull-kubernetes-node-e2e-containerd

liggitt commented on cmd/kubelet/app/server.go:
```diff
@@ -841,15 +841,26 @@ func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, nod
 		// we set exitAfter to five minutes because we use this client configuration to request new certs - if we are unable
 		// to request new certs, we will be unable to continue normal operation. Exiting the process allows a wrapper
 		// or the bootstrapping credentials to potentially lay down new initial config.
-		closeAllConns, err := kubeletcertificate.UpdateTransport(wait.NeverStop, transportConfig, clientCertificateManager, 5*time.Minute)
```
I think it would be clearer to keep this var `closeAllConns` (since that's actually what this one does), and set up a separate `var onHeartbeatFailure func()` below which we set either to `closeAllConns` or to `func() { utilnet.CloseIdleConnectionsFor(transportConfig.Transport) }`

aojea replied:
yep, that is right, let me fix it

ryanzhang-oss commented:
Doesn't this need any test cases?

aojea commented:
> Doesn't this need any test cases?

Yes, but I´ve already added those test cases with all the permutations of http protocols and network problems in https://github.com/kubernetes/kubernetes/pull/104844 , this is just extending the same solution to the other path that was generating the `onHeartbeatFailure` hook

ryanzhang-oss commented:
> > Doesn't this need any test cases?
> 
> Yes, but I´ve already added those test cases with all the permutations of http protocols and network problems in #104844 , this is just extending the same solution to the other path that was generating the `onHeartbeatFailure` hook

I am not sure if those unit tests clearly indicate the benefit of this PR but I don't know the whole picture.

aojea commented:
https://github.com/kubernetes/kubernetes/pull/104844/commits/b9d865a8185b62d83e9ff81b0e3499a26ac6960d

they show that the new function is able to recover from a situation that the TCP connection is broken but the endpoint is not aware, without requiring with close the whole connection, just forcing the client to try a new connection

ryanzhang-oss commented:
> [b9d865a](https://github.com/kubernetes/kubernetes/commit/b9d865a8185b62d83e9ff81b0e3499a26ac6960d)
> 
> they show that the new function is able to recover from a situation that the TCP connection is broken but the endpoint is not aware, without requiring with close the whole connection, just forcing the client to try a new connection

I think the rule of thumb is that we need at least one test that fails before the fix and passes after. I am not sure if we can create a test case like that here.

wojtek-t commented:
/lgtm
/approve

Thanks!

JohnRusk commented:
@aojea Awesome to see this merged. Thank you!  Are there any plans to create cherry pick PRs to get this into 1.23 (and maybe 1.22)?  That could be helpful for users currently suffering from the issue but not ready for a major version upgrade.

liggitt commented:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/cherry-picks.md#what-kind-of-prs-are-good-for-cherry-picks

since this isn't fixing a regression from those versions, and I'm pretty skeptical it rises to the level of critical bug fix, I wouldn't really expect a backport of this

jackfrancis commented:
@liggitt The side-effect of certain cluster behaviors at scale seems to meet the "Panic, crash, hang" criterion in the cherry-pick definition. This change will ameliorate those apiserver degradation scenarios. Are we in disagreement about that? Or is the actual cherry-pick process itself non-trivial (composing a change as multiple PRs out-of-sequence, stuff like that)

liggitt commented:
While we don't anticipate issues with this fix (otherwise we wouldn't have merged it), backporting exposes release branches to unanticipated issues. Since this is a historically fragile area, I'd be extremely cautious about taking back a fix here for anything other than a regression in one of those releases

JohnRusk commented:
> anything other than a regression in one of those releases

FYI, my understanding is that this issue is a regression, but that the regression happened several releases ago. (e.g. in 1.18 or something. I haven't checked exactly).  If that's corrrect does it change anything in your reply @liggitt?  (I imagine not, but I'm just checking :-))

jackfrancis commented:
Understand the practical realities at play here, thx for clarifying @liggitt

liggitt commented:
> FYI, my understanding is that this issue is a regression, but that the regression happened several releases ago.

https://github.com/kubernetes/kubernetes/pull/63492 merged in 1.11 and was picked back to 1.8.x, so this behavior has existed since then. Choosing between an edge case that can result in a crash at scale and an edge case that results in a silent and ~unrecoverable hang of nodes is not a super clear choice. I'd still lean against backporting this.

aojea commented:
I agree with Jordan judgement, however, in case of backport , this can only go maximum to 1.23, since this depends on https://github.com/kubernetes/kubernetes/pull/104844/commits/b9d865a8185b62d83e9ff81b0e3499a26ac6960d and that  is clearly not backportable

JohnRusk commented:
Thanks for the clarification guys.  I understand your reasoning.

djsly commented:
So this will be in 1.25 only ?  We are currently on 1.21 and affected  once a week/two weeks in one of our clusters at random.

ehashman commented:
@djsly this is in 1.24 and will be backported to 1.23.

ehashman commented:
/priority important-soon
